### PR TITLE
feat(query): Support query result cache system table and table function

### PR DESCRIPTION
**Pull request by ariesdevil**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR
- [x] query result cache system table
- [x] last_query_id context function
- [x] tests

Tracking in #10010 .

### CODE REVIEW

File: src/query/sql/src/planner/semantic/type_check.rs [hunk 1]
```diff
@@ -1642,6 +1643,44 @@ impl<'a> TypeChecker<'a> {
                 )
             }
 
+            ("last_query_id", args) => {
+                let index = if args.len() != 1 {
```
Comment (sundy-li) [inline src/query/sql/src/planner/semantic/type_check.rs:1647]: we can folder it into a constant `expr`.

**Comment by ariesdevil** [comment]
query result cache table function needs to generate schema dynamically, which will done in next PR.

File: src/query/storages/result_cache/src/write/writer.rs [hunk 2]
```diff
@@ -69,6 +69,7 @@ impl ResultCacheWriter {
     }
 
     pub fn num_rows(&self) -> usize {
+        println!("number rows: {}", self.num_rows);
```
Comment (sundy-li) [inline src/query/storages/result_cache/src/write/writer.rs:72]: extra print

File: src/query/service/src/sessions/query_ctx_shared.rs [hunk 3]
```diff
@@ -336,6 +336,14 @@ impl QueryContextShared {
     }
 }
 
+impl Drop for QueryContextShared {
+    fn drop(&mut self) {
+        self.session
+            .session_ctx
+            .update_query_ids_results(self.init_query_id.read().clone(), None)
+    }
+}
```
Comment (RinChanNOWWW) [inline src/query/service/src/sessions/query_ctx_shared.rs:345]: Why need to do this?

File: src/query/sql/src/planner/semantic/type_check.rs [hunk 4]
```diff
@@ -1642,6 +1643,57 @@ impl<'a> TypeChecker<'a> {
                 )
             }
 
+            ("last_query_id", args) => {
```
Comment (RinChanNOWWW) [inline src/query/sql/src/planner/semantic/type_check.rs:1646]: Can add some comments to describe the rewrite results. Like other rewriteable functions.

File: src/query/storages/system/src/query_cache_table.rs [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    pub fn create(table_id: u64) -> Arc<dyn Table> {
+        let schema = TableSchemaRefExt::create(vec![
+            TableField::new("sql", TableDataType::String),
+            TableField::new("result_size", TableDataType::Number(NumberDataType::UInt64)),
+            TableField::new("num_rows", TableDataType::Number(NumberDataType::UInt64)),
+            TableField::new(
+                "partitions_sha",
+                TableDataType::Array(Box::new(TableDataType::String)),
+            ),
+            TableField::new("locations", TableDataType::String),
```
Comment (RinChanNOWWW) [inline src/query/storages/system/src/query_cache_table.rs:103]: Each cache only has one location now, so I think the table field name should be 'location'.

**Review by sundy-li — APPROVED** [review]

**Review by RinChanNOWWW — APPROVED** [review]

**Pull request merged**